rfcs/60-network-error.md

```diff
@@ -0,0 +1,204 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2021-05-11
+---
+
+# AOS-60: Network Error and Retryable Server-side Errors
+
+## Background
+
+We can classify all errors by the root cause of error:
+
+- User-side: incapability, wrong usage, pass an append object into multipart operations (like `failed to send a request to the storage system`)
+- Network-side: Network related error (like `failed to receive a response from the storage system`)
+  - explicit error: connection refuse, connection reset
+  - implicit error: timeout
+- Service-side: not exist, permission denied, already exists... (like `received an error response from the storage system`)
+
+All the errors will go into `formatError`, and be converted into either a defined error or an unexpected error. Currently, we have only defined some user-side errors and server-side errors. 
+
+For errors which are returned by SDK, but not belong to the SDK's error type, they are either belong to user-side or network-side. 
+
+```go
+func formatError(err error) error {
+	if _, ok := err.(services.AosError); ok {
+		// client-side
+		return err 
+	}
+
+	var e *qserror.QingStorError
+	if !errors.As(err, &e) {
+		// client-side or network-side
+		return fmt.Errorf("%w: %v", services.ErrUnexpected, err)
+	}
+
+	// server-side
+	switch e.Code {
+	case "permission_denied":
+		return fmt.Errorf("%w: %v", services.ErrPermissionDenied, e)
+	case "object_not_exists":
+		return fmt.Errorf("%w: %v", services.ErrObjectNotExist, e)
+	default:
+		return fmt.Errorf("%w: %v", services.ErrUnexpected, err)
+	}
+}
+```
+
+Network errors are common and should be expected. Besides, there's a common need to retry a network error. So we should provide network error codes. 
+
+It will be better if a good retry mechanism can be built on top of network errors.
+
+## Proposal
+
+So I propose to add the following global error codes:
+
+Network-side: 
+
+- `ErrNetworkTemporary`: context.canceled, connection timeout, no response for some time 
+
+And use the following function to judge and convert.
+
+```go 
+func IsErrNetworkTemporary(err error) bool {
```

Comment:
How about add a `NetworkError` struct and let the service provider decide whether this error is a NetworkError.

If so, we can:

- Add a helper function in go-storage to simplify the service provider's work
- Add a new interface `Temporary` (just example), any error implement this interface can be retried safely.
- Add more specific error in NetworkError like timeout

After all this work, users can:

- Know if this error retryable
- Know if this error a network error

Reply:
> How about add a `NetworkError` struct

With `ErrNetworkTemporary` as an error code, services can always create detailed network error structs. What's the difference?

Reply:
First, I hope to provide a judgement like `IsRetryable` or `IsTemporary` (for user), then a retry mechanism utility may also be built beside this judgement. Whether to use an interface like `Temporary` is an implementation detail.

And due to GSP-47, we should have some error codes for network errors, possibilities are:
- Only `ErrNetwork`
- `ErrNetworkTemporary` + others (not sure)
- Only detailed codes `ErrConnectionRet`

The benefit of the second way is that no additional effort to implement `IsRetryable` judgement is needed. (not a very big benefit🤔)

Reply:
The problem I think about is: What's the relation between `NetworkError` / `TemporaryError` / `RetryableError`?

To me:

`ErrConnectionReset` looks like a specific error. The `NetworkError` is its category. And `Temporary` is a property of it.

So here are combinations:

- NetworkTemporary
- NetworkPermenent
- AnotherCategoryTemporary
- AnotherCategoryPermenent
- ...

It's not LGTM that we need to maintain so many error codes.

Further, not all network error is temporary:

- connection reset could be temporary
- DNS resolved failed could be permanent

These why I think the error code `ErrNetworkTemporary` is not a good idea.

To `IsErrNetworkTemporary` or `IsRetryable`, I'm afraid it's hard to maintain.

> Provide mechanisms rather than strategies

I prefer to let services handle all errors instead of checking them by the user. That means, all errors that returned by services or go-storage SHOULD not be raw errors like `io.ErrUnexpectedEOF` or `url.Error`. They all SHOULD be wrapped by `formatError` or check functions provided by us.

Reply:
> What's the relation between NetworkError / TemporaryError / RetryableError
This is also what I was considering. And I agree with your statement.

> Further, not all network error is temporary:
Of course. Otherwise, I would simply give an `ErrNetwork` instead of `ErrNetworkTemporary`. Permanent network errors are not specified here.

Reply:
> I prefer to let services handle all errors instead of checking them by the user. That means, all errors that returned by services or go-storage SHOULD not be raw errors like io.ErrUnexpectedEOF or url.Error. They all SHOULD be wrapped by formatError or check functions provided by us.

I completely agree. But I think this is irrelevant to our current discussion. (We did not violate the rules.)

Reply:
> These why I think the error code ErrNetworkTemporary is not a good idea.
Actually I also thought it not a good idea, but didn't come up with a better one.

Reply:
How about `ErrNetwork` error code + `IsRetryableError` interface?

> I use retryable instead of temporary for
> - retry is an action that users can take, they can decide to retry it or not.
> - temporary is a state which could be changed and unpredictable. For example, it's hard to decide whether DNS resolved failed is temporary.

Reply:
You mean: `ErrNetwork` error code + structs like `ConnectionResetError`? 

How detailed errors should we provide? There may be a long list: ConnectionReset, ConnectionRefused, ...

We should consider the list and how to easily identify and transform them in `formatError` (like `IsErrNetworkTemporary` in my original proposal).

Reply:
> How detailed errors should we provide? There may be a long list: ConnectionReset, ConnectionRefused, ...

I thought users who use `go-storage` may not care about detailed network error?

We can define following:

- `ErrNetworkUnexpected`
- `ErrNetworkTimeout`
- ...

This list should be extended one by one and only added errors that user care about like `timeout`.

> We should consider the list and how to easily identify and transform them in formatError (like IsErrNetworkTemporary in my original proposal).

I prefer to let the service decide them. Maybe we can extract those helper functions in a new package later, but they are out of the scope of this GSP.